src/audio/dai-zephyr.c

```diff
@@ -0,0 +1,1466 @@
+		if (!group->group_id)
+			group->group_id = group_id;
+
+		group->num_dais++;
+	} else {
+		tr_err(&dai_comp_tr, "dai_group_get(): failed to get group_id %u",
+		       group_id);
+	}
+
+	return group;
+}
+
+void dai_group_put(struct dai_group *group)
+{
+        group->num_dais--;
+
+        /* Mark as unused if there are no more DAIs in this group */
+        if (!group->num_dais)
+                group->group_id = 0;
+}
+
+/* Assign DAI to a group */
+int dai_assign_group(struct comp_dev *dev, uint32_t group_id)
+{
+	struct dai_data *dd = comp_get_drvdata(dev);
+
+	if (dd->group) {
+		if (dd->group->group_id != group_id) {
+			comp_err(dev, "dai_assign_group(), DAI already in group %d, requested %d",
+				 dd->group->group_id, group_id);
+			return -EINVAL;
+		}
+
+		/* No need to re-assign to the same group, do nothing */
+		return 0;
+	}
+
+	dd->group = dai_group_get(group_id, DAI_CREAT);
+	if (!dd->group) {
+		comp_err(dev, "dai_assign_group(), failed to assign group %d",
+			 group_id);
+		return -EINVAL;
+	}
+
+	comp_dbg(dev, "dai_assign_group(), group %d num %d",
+		 group_id, dd->group->num_dais);
+
+	/* Register for the atomic trigger event */
+	notifier_register(dev, dd->group, NOTIFIER_ID_DAI_TRIGGER,
+			  dai_atomic_trigger, 0);
+
+	return 0;
+}
+
+/* called from ipc/ipc3/handler.c and some platform.c files */
+struct dai *dai_get(uint32_t type, uint32_t index, uint32_t flags)
+{
+	const struct dai_type_info *dti;
+	struct dai *d;
+
+	tr_info(&dai_comp_tr, "log_ssp_data get_dai_zephyr index: %u", index);
+
+	dti = dai_find_type(type);
+	if (!dti)
+		return NULL; /* type not found */
+
+	for (d = dti->dai_array; d < dti->dai_array + dti->num_dais; d++) {
+		if (d->index == index)
+			return dai_probe(d->drv) ? NULL : d;
+	}
+
+	return NULL;
+}
+
+/* called from src/ipc/ipc3/handler.c */
+void dai_put(struct dai *dai)
+{
+	int ret;
+
+	ret = dai_remove(dai->drv);
+	if (ret < 0) {
+		tr_err(&dai_comp_tr, "dai_put_zephyr: index %d failed ret = %d",
+		       dai->index, ret);
+	}
+}
+
+/* called from all multiple platform.c files */
+int dai_init(struct sof *sof)
+{
+	struct dai *dai;
+	int i;
+
+#if CONFIG_INTEL_SSP
+	struct dai_type_info *ssp = NULL;
+	char ssp_name[] = "SSP_X";
+#endif
+#if CONFIG_INTEL_ALH
+	struct dai_type_info *alh = NULL;
+#endif
+#if CONFIG_INTEL_DMIC
+	struct dai_type_info *dmic = NULL;
+	char dmic_name[] = "DMICX";
+#endif
+
+	for (i = 0; i < ARRAY_SIZE(dti); i++)
+		switch (dti[i].type) {
+#if CONFIG_INTEL_SSP
+		case SOF_DAI_INTEL_SSP:
+			ssp = dti + i;
+			break;
+#endif
+#if CONFIG_INTEL_ALH
+		case SOF_DAI_INTEL_ALH:
+			alh = dti + i;
+			break;
+#endif
+#if CONFIG_INTEL_DMIC
+		case SOF_DAI_INTEL_DMIC:
+			dmic = dti + i;
+			break;
+#endif
+		}
+
+#if CONFIG_INTEL_SSP
+	dai = ssp->dai_array;
+	ssp->num_dais = 0;
+
+	/* init ssp */
+	for (i = 0; i < SOF_MAX_SSP_DEVICE; i++) {
+		/* name varied by zephyr device tree */
+		ssp_name[4] = (uint8_t)((i % 10UL) + '0');
```

Comment:
Why are not using a table for this ? I've a previous comment on this (but maybe been resolved ?)